src/globus_cli/parsing/param_classes.py

```diff
@@ -1,14 +1,50 @@
+from __future__ import annotations
+
+import typing as t
+
 import click
 
+C = t.TypeVar("C", bound=t.Union[click.BaseCommand, t.Callable])
+
+
+class _Sentinel:
+    pass
+
+
+_SENTINEL = _Sentinel()
+
 
-class OneUseOption(click.Option):
+class AnnotatedOption(click.Option):
+    def __init__(
+        self,
+        *args: t.Any,
+        type_annotation: type | _Sentinel = _SENTINEL,
```

Comment:
Can this value validly be `None`?

Confused why we'd introduce the `_SENTINEL` "i didn't specify it" singleton.

Reply:
`None` is valid in annotations, so I believe it can be validly expressed as `None`. Such an option would not necessarily be useless, though it would be a little unusual.

For example:
```python
def callback(ctx, param, value):
    warnings.warn("got deprecated option --foo")
    return None

@click.option("--foo", cls=AnnotatedOption, type_annotation=None, hidden=True, callback=callback)
def myfunc(*, foo: None) -> None:
    ...
```

If `--foo` is deprecated, it's not _necessary_ to make its type `None`. The following nearly-equivalent thing is also fine:
```python
def callback(ctx, param, value):
    warnings.warn("got deprecated option --foo")
    return value

@click.option("--foo", hidden=True, callback=callback)
def myfunc(*, foo: str | None) -> None:
    ...
```

But since this is meant to be at the core of the click type checker, I wanted to account for the possibility.